marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/impl/entity-lib.sjs

```diff
@@ -0,0 +1,133 @@
+/**
+ Copyright 2012-2019 MarkLogic Corporation
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+ http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ */
+'use strict';
+
+const sem = require("/MarkLogic/semantics.xqy");
+const semPrefixes = {es: 'http://marklogic.com/entity-services#'};
+
+/**
+ * @return an array of strings, one for each EntityType
+ */
+function findEntityIRIs() {
+  return cts.triples(null,
+    sem.iri('http://www.w3.org/1999/02/22-rdf-syntax-ns#type'),
+    sem.iri('http://marklogic.com/entity-services#EntityType'), '='
+  ).toArray().map(triple => sem.tripleSubject(triple).toString());
+}
+
+/**
+ * @return a map object where each key is an entity IRI, and the value of each key is the corresponding entity type
+ */
+function findEntityTypesAsMap() {
+  const map = {};
+  for (var doc of cts.search(cts.collectionQuery("http://marklogic.com/entity-services/models"))) {
+    Object.assign(map, convertModelToEntityIRIMap(doc.toObject()));
+  }
+  return map;
+}
+
+/**
+ * @param entityIRI string or a sem.iri
+ * @return the matching model document, or null if one is not found
+ */
+function findModelForEntityIRI(entityIRI) {
+  return fn.head(cts.search(
+    cts.andQuery([
+      cts.collectionQuery('http://marklogic.com/entity-services/models'),
+      cts.tripleRangeQuery(sem.iri(entityIRI), sem.curieExpand("rdf:type"), sem.curieExpand("es:EntityType", semPrefixes), "=")
+    ])));
+}
+
+/**
+ * @param entityIRI string or a sem.iri
+ * @return null if a model can't be found matching the given IRI, or if a model is found but there's no an entity type
+ * with a title matching the entity title in the IRI. Otherwise, the entity type from the definitions array in the model
+ * is returned.
+ */
+function findEntityType(entityIRI) {
+  const modelDoc = findModelForEntityIRI(entityIRI);
+  if (!modelDoc) {
+    return null;
+  }
+
+  const parts = getEntityIRIParts(entityIRI);
+  const entityTitle = parts.entityTitle;
+  return modelDoc.toObject().definitions[entityTitle];
+}
+
+/**
+ * @param entityIRI sem.iri or string
+ * @return {{entityTitle: string, baseUri: string, modelTitle: string, version: string}}
+ */
+function getEntityIRIParts(entityIRI) {
```

Comment:
I don't think we can do much better than this approach with what we have available form entity services today, but I think raises additional needs we have the logical model. There is no validation from the ES libraries that prevent the Entity Model TItle or the Entity Type Title from having slashes in them, so we can technically have an IRI that looks like: `http://marklogic.com/example/models/ProductDetail/d-0.0.1/ProductDetail/d` I guess perhaps we could drop into the Entity Model Descriptor document itself and get the baseUri and then use the triples to get the rest of the information?